src/Compiler/Microsoft.AspNetCore.Razor.Language/legacyTest/Microsoft.AspNetCore.Razor.Language.Legacy.Test.csproj

```diff
@@ -21,6 +21,9 @@
 
     <PackageReference Include="Newtonsoft.Json" />
     <PackageReference Include="Moq" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" VersionOverride="$(MicrosoftCodeAnalysisCSharpWorkspacesPackageVersion)" />
```

Comment:
Just for my own understanding, why are these neccessary?

Reply:
I had the same question. Does arcade do something to these packages? Because it seems like the override version would otherwise always match the real version.

Reply:
I assume it's because these projects depend on these packages implicitly, and if we didn't set the version explicitly here, they'd be set to the version in the _root_ project, not the overridden version in the compiler projects.

Reply:
What is going on here is that there are three sources of two sources of references for `MS.CA.CSharp` (and the other projects):

1. The project ref to Microsoft.CodeAnalysis.Razor.Compiler.csproj which uses version 4.9.2
2. The project ref to Microsoft.AspNetCore.Razor.Test.Common.csproj which uses version 4.12.0-1.24379.11

Standard MSBuild logic will say the 4.12.0-1.24379.11 reference is higher therefore lets use that one. But in this case we're using central transitive pinning which means that the version in `Directory.Packages.props` wins over standard MSBuild logic. For this project it uses `src\Compiler\Directory.Packages.props` which has `MS.CA.CSharp` at 4.9.2. Without this `VersionOverride` we'd end up with un-runnable binaries.

The good news here is that an _essential_ feature of central transitive package pinning is that it will warn aggressively on any silent downgrades like this. That is _essential_ because central transitive pinning is somewhat spooky action at a distance. It's saying any transitive dependency that matches an entry in `Directory.Packages.props` will be done at the version specified in that file. Hence it aggressively warns if this ever results in a silent downgrade because that is almost certainly what you did not want. 

That is exactly what happened here and why I was lead to make the correct override notations.

Reply:
Thank you for the explanation. I didn't realise there still was a Directory.Packages.props in the Compiler folder, so I couldn't think of a place where a different version could possibly come from.

Reply:
Yeah it's still there because we haven't fully gotten the torn state solution into the .NET SDK. Once we get to say .NET 9 SDK RC1 we don't have to pin the compiler backwards any more and we can start undoing this nest of weirdness.